### DISCUSSION TITLE
How to get `data.meta` from markdown snippet code after using `rehype-raw`?

### DISCUSSION THREAD

**Discussion by AhmedBHameed** [remarkjs/remark#1026] [Q&A] [answered]
`react-markdown` is taking me here to discuss.

Hi there,

I'm using markdown to render post content using `react-markdown`. I encountered with snippet code that using some metadata where needed to clarify the code

Something like this:

```
` ` `tsx {1, 3-10} [src/components/SupportLink/SupportLink.tsx]

// the code snippet here

 ` ` `
```

More to that, there is a custom html wher needed to be rendered in the page. I used `rehype-raw` to do the job **but,** for a reason, I can't see meta data anymore?

I found this answer here https://github.com/syntax-tree/hast-util-raw/issues/13 but I'm not sure how to configure the plugin?

A minimal reproduction can be found here

https://github.com/AhmedBHameed/next-rehype-raw-issue

Images from this reproduction:

**Before using `rehype-raw`**

![issue](https://user-images.githubusercontent.com/25675361/184009483-8046308e-f8c6-4335-ad45-098aa0e4dd14.jpg)

**After using `rehype-raw`**

![issue0](https://user-images.githubusercontent.com/25675361/184009463-f697655b-ad09-4430-9dc8-2a353403ac33.jpg)

I decided to open a discussion since https://github.com/rehypejs/rehype-raw/issues/9#issuecomment-1211166088

Please note that I searched dicsussions before posting. I found something related to `unified` but not sure if this is what I need to do?
So If you know this is duplicate please point me and execuse my inattention

Thank you in advance.

**Comment by ChristianMurphy** [comment, marked as answer]
Consider copying the internal hast `data` attribute over to an actual HTML data attribute like https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes

**Reply by AhmedBHameed** [reply]
@ChristianMurphy 

Hmm, TBH did not think about that. Seems to me possible, but have to test it tho.

In general, `rehype-raw` is wiping the meta data for a reason, and since wooorm mentioned https://github.com/rehypejs/rehype-raw/issues/9#issuecomment-1211166088 which take me to this issue https://github.com/syntax-tree/hast-util-raw/issues/13 but nothing usefull there?

> You can handle data fields in a plugin beforehand. Please open a discussion if you have questions about that.

That is why this discussion created.

It might be a cleaner way to keep writing markdown format instead of putting the html tag for every code snippet.
Thank you for the idea.

**Reply by wooorm** [reply]
> > You can handle data fields in a plugin beforehand.

That comment implies, in your practical case, indeed exactly what Christian says: handle them before by using the information and adding it as actual attributes.

> Seems to me possible, but have to test it tho.

It’s definitely possible :)

---

Taking a step back: Why do you even use `rehype-raw`? Like because there is HTML too in the markdown, right?
So: how could the same thing as your fancy code block, with markdown, be created with HTML?
By putting that information on the elements. As attributes!

**Reply by AhmedBHameed** [reply]
The reason of using `rehype-raw` to handle custom HTML that is not "easly" supported via markdown (I might be wrong tho) like:

```html
<video width="100%" controls autoplay loop crossorigin="anonymous">
  <source src="URL" type="video/webm">
</video>
```
I am aware that markdown is converted to HTML afterall, however, it does make difference IMO:

- It is cleaner to store markdown mostly in DB.
- I try as much as possible to avoid polluting the markdown with HTML to avoid any future break changes in case find something better than `react-markdown`
- Since I want to write technical articles, adding markdown snippet code syntax regullarly is much easier than adding `pre` tag.

Don't you agree ?

Would you mind to clarify why `rehype-raw` intentially removing meta data? Can I configure it somehow with `react-markdown` ?

**Reply by wooorm** [reply]
Your response here has nothing to do with the problem at hand?

I already explained why metadata is removed: https://github.com/syntax-tree/hast-util-raw/issues/13#issuecomment-912451531.

---

I believe you’re missing what Christian and I are trying to tell you.
We are **not** suggesting that you write HTML manually with dataset attributes.
We are suggesting that you **write a plugin, that is used before `rehype-raw`, which turns the hidden data fields into actual dataset properties**.

**Reply by AhmedBHameed** [reply]
To me that line

> It is intended that the tree is changed and unrelated data lost, per the readme.

Does not clarify why? Please be more collaborative with your answers. I think metadata is related so seems to me wiping them is not right?


> Your response here has nothing to do with the problem at hand?

It does related to your question above and that was my answer to express my scenario.


Ok I might missundetand you both.

> We are suggesting that you write a plugin, that is used before rehype-raw, which turns the hidden data fields into actual dataset properties.

That seems good point to start. I will search about it but in case you have links related please share it here?

**Reply by wooorm** [reply]
Read https://github.com/unifiedjs/unified up until the API section, then look at a few guides on https://unifiedjs.com, finally look at a couple of existing plugins that do things similar to what you want to do.

**Reply by AhmedBHameed** [reply]
Nice! Will do. Thanks @wooorm  for your patience and the guidness.
